### ISSUE TITLE
Training Results do not match Test Results [Pix2Pix]

### ISSUE THREAD

**Issue by awhelan-school** [junyanz/pytorch-CycleGAN-and-pix2pix#1393]
I am training a conditional model to generate logos. However, the results outputting during training are much better than when I try to do inference on a single image. For example, 

Real A
![epoch021_real_A](https://user-images.githubusercontent.com/27755651/157329816-aca745ff-4e37-4b4f-a2f9-32f36b6e60c5.png)

Real B
![epoch021_real_B](https://user-images.githubusercontent.com/27755651/157329718-56c812dc-134e-477b-9457-c4f3f411c02c.png)

Fake  B
![epoch021_fake_B](https://user-images.githubusercontent.com/27755651/157329839-fd4a1349-ff07-4ab4-999e-e1838ed54bc1.png)

Which I think is really good. However, when I run test I get 

![output_fake](https://user-images.githubusercontent.com/27755651/157329993-c577258b-d974-4d59-b432-6c407407a940.png)

One thing I noticed that occurs across all test images is that the colors seem to "bleed" across the edges. Could this have something to do with how my test input image is saved or do I need to do some sort or pre-processing step to make the test results more accurate to the training results.

I am also getting this warning when I run test.py. Not sure if it is relevant.
UserWarning: Argument interpolation should be of type InterpolationMode instead of int. Please, use InterpolationMode enum.

Thanks in advance

**Comment by junyanz** [comment]
It could be caused by several potential reasons: 
(1) Your test and training scripts use different options: You can run the test.py with the same input image as used during the training time. You should expect to get results of similar quality. 
(2) Your model might overfit the training set if (1) is resolved but your model does not work for test inputs. You may consider collecting more images or use more data augmentation. 
(3) Bleeding artifacts: You might get bleeding artifacts when your input edges are not closed (due to data pre-processing and augmentation). You recommend that your data augmentation should not create open edges.

**Comment by awhelan-school** [comment]
I am using 100K images so I don't think the amount of data is the issue. I also tried to make sure the options are the same for test and train. If I am not mistaken the only options that are mentioned in the FAQ that we must verify are the same are 
--norm=batch, --netG=unet_256, batch_size=16. I also diffed them to make sure they are the same.

batch_size: 16 16
checkpoints_dir: ./checkpoints ./checkpoints
crop_size: 256 256
dataroot: Test/ dataset.p2p/C
dataset_mode: single aligned
direction: AtoB AtoB
display_winsize: 256 256
epoch: latest latest
gpu_ids: 0 0
init_gain: 0.02 0.02
init_type: normal normal
input_nc: 3 3
isTrain: False True
load_iter: 0 0
load_size: 256 286
max_dataset_size: inf inf
model: test pix2pix
n_layers_D: 3 3
name: logoP2P logoP2P_V2
ndf: 64 64
netD: basic basic
netG: unet_256 unet_256
ngf: 64 64
no_dropout: False False
no_flip: False False
norm: batch batch
num_threads: 4 4
output_nc: 3 3
phase: test train
preprocess: resize_and_crop resize_and_crop
serial_batches: False False
use_wandb: False False
verbose: False False

The only option that is different is dataset_mode which is single in test and aligned in train which is to be expected or am I wrong?

 Could it be that when you output the fake result at a specific epoch that result is metric wise the best one deemed by the discriminator for that epoch and at a later epoch it may not hold true. Because, the result illustrated is from epoch 21 and I am using the model at epoch 30.

**Comment by ShiLayne** [comment]
I have the same problem in [CycleGAN]. The middle results (200epoch/400epoch in total) in training are very good, but when I use the same image in test mode (with just A to B generator and "single_dataset" and not modify other setting), it shows a huge difference.

**Comment by ShiLayne** [comment]
> 

Sry for I may resolve my problem. I use batch norm in CycleGAN for test before and when eval it runs model.eval(). Normalization does greatly affect the generated results. After I delete model.eval(), the test results returned to good training result. Hope it will be helpful to you.

**Comment by junyanz** [comment]
Yes, make sure that you use the same normalization during training and test. 
@awhelan-school You can try to use the test code with epoch 21.

**Comment by awhelan-school** [comment]
Thanks for the replies. I think I understand. If I want to be doing testing with only a single image I should have trained the network using instance normalization. I tried providing multiple test images and it seems the results are much closer to the ones reported during training since I trained with batch normalization. Anyway thanks for the help and appreciate the code I think its a great way to demo how we can do style transfer.

**Comment by q671383789** [comment]
I'm confused , while testing Why should we use train mode?
Because if you use train mode, the parameter of bn will be changed, and drop-out will be use.
But if you want to get the same result with the same input, the bn shouldnt be changed .

**Comment by junyanz** [comment]
It should not affect a lot if you use instance normalization. Things might get interesting as we use bn=1 with batchnorm in many pix2pix experiments. In that case, you may want to keep train mode during test time. Feel free to try both.

**Comment by q671383789** [comment]
> It should not affect a lot if you use instance normalization. Things might get interesting as we use bn=1 with batchnorm in many pix2pix experiments. In that case, you may want to keep train mode during test time. Feel free to try both.

Okay, I have tried. It seems like in my task it seems like instance norm is less perform than bn>1, and I have to use eval mode during inference otherwise the output when I testing will be worse
